minio/helpers.py

```diff
@@ -839,8 +839,10 @@ def run(self):
                     self._results_queue.put(result)
                 except Exception as ex:  # pylint: disable=broad-except
                     self._exceptions_queue.put(ex)
-                finally:
-                    cleanup_func()
+
+            # cleanup_func represents semaphore.release() to manage thread pool capacity,
+            # it should be called unconditionally to avoid semaphore leaks and deadlocks.
+            cleanup_func()
```

Comment:
I do not understand the fix. How does it different from `finally`. If it is `python` bug, report `python`.

Reply:
see the minimal reproduction code, it would block forever @balamurugana

Reply:
As the reproducer is generic python code, why is it minio-py issue?

Reply:
fixing my PR today I lurked to this one out of curiosity, and clearly @detailyang is right

> As the reproducer is generic python code, why is it minio-py issue?

I think because minio-py is using this reproducer in it's code and reproducer has bug :smile: 

---

You can leave "cleanup_func()" in finally block as is, it doesn't matter (but with the line added, it's redundant). The bug is that if **code doesn't go inside** `if self._exceptions_queue.empty()` if statement block, then semaphore is never released (it never reaches finally block)

In other words, it's directly related to using acquire in add_task, and releasing in workers. There is a comment `# Mark this task as done, whether an exception happened or not` and note that marking task as done should also always release semaphore

Reply:
@balamurugana Please take another look at this; the bug is actually quite serious and affects the majority of versions. Many thanks:)